File: selenium-e2e/src/test/java/dev/dsbon/realworld/support/SeedUsers.java
package dev.dsbon.realworld.support;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

/**
 * Reads {@code shared/seed-users.json} — the same fixture the Cypress and
 * Playwright suites consume, so all three frameworks drive identical accounts.
 *
 * <p>The file lives outside this Maven module, which is deliberate: duplicating
 * the fixture per framework is how three suites quietly drift apart until two of
 * them are testing an account that no longer exists.
 */
public final class SeedUsers {

  private static final Path FIXTURE = locateFixture();
  private static final SeedFile SEED = load();

  private SeedUsers() {}

  public static String defaultPassword() {
    return SEED.defaultPassword();
  }

  /** The user every test signs in as unless it says otherwise. */
  public static SeedUser primary() {
    return byRole("primary");
  }

  /** A second, distinct account — the counterparty in payment tests. */
  public static SeedUser contact() {
    return byRole("contact");
  }

  private static SeedUser byRole(String role) {
    return SEED.users().stream()
        .filter(u -> role.equals(u.role()))
        .findFirst()
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "No seed user with role \"%s\" in %s. The app may have been re-seeded with "
                            .formatted(role, FIXTURE)
                        + "\"yarn db:seed\" (random data) instead of \"yarn db:seed:dev\"."));
  }

  /**
   * Walk up from the working directory until {@code shared/seed-users.json} turns
   * up. Surefire runs with the module directory as CWD, but IDEs frequently use
   * the repository root instead — searching upward makes both work without a
   * configuration flag.
   */
  private static Path locateFixture() {
    Path dir = Path.of("").toAbsolutePath();
    for (int i = 0; i < 5 && dir != null; i++, dir = dir.getParent()) {
      Path candidate = dir.resolve("shared").resolve("seed-users.json");
      if (Files.isRegularFile(candidate)) {
        return candidate;
      }
    }
    throw new IllegalStateException(
        "Could not find shared/seed-users.json above " + Path.of("").toAbsolutePath());
  }

  private static SeedFile load() {
    try {
      return new ObjectMapper().readValue(Files.readString(FIXTURE), SeedFile.class);
    } catch (IOException e) {
      throw new UncheckedIOException("Failed to read " + FIXTURE, e);
    }
  }

  @JsonIgnoreProperties(ignoreUnknown = true)
  private record SeedFile(String defaultPassword, List<SeedUser> users) {}

  @JsonIgnoreProperties(ignoreUnknown = true)
  public record SeedUser(String id, String username, String firstName, String lastName, String role) {}
}

File: playwright-e2e/fixtures/seed-users.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

export interface SeedUser {
  id: string;
  username: string;
  firstName: string;
  lastName: string;
  role: "primary" | "contact" | "spare";
}

interface SeedFile {
  defaultPassword: string;
  users: SeedUser[];
}

/**
 * The same `shared/seed-users.json` the Cypress and Selenium suites read, so all
 * three frameworks drive the same accounts and a change to the app's seed data
 * is a one-file fix.
 */
const seed: SeedFile = JSON.parse(
  readFileSync(fileURLToPath(new URL("../../shared/seed-users.json", import.meta.url)), "utf8"),
);

export const DEFAULT_PASSWORD = seed.defaultPassword;

export function userByRole(role: SeedUser["role"]): SeedUser {
  const match = seed.users.find((u) => u.role === role);
  if (!match) {
    throw new Error(
      `No seed user with role "${role}" in shared/seed-users.json. The app may have been ` +
        `re-seeded with "yarn db:seed" (random data) instead of "yarn db:seed:dev".`,
    );
  }
  return match;
}

export const PRIMARY_USER = userByRole("primary");
export const CONTACT_USER = userByRole("contact");

/** Storage state written by the setup project and reused by every test. */
export const STORAGE_STATE = fileURLToPath(new URL("../.auth/primary-user.json", import.meta.url));
